rust/src/storage/s3/store.rs

```diff
@@ -0,0 +1,658 @@
+            Some(keys)
+        } else {
+            None
+        }
+    }
+
+    // Delete the objects of the given bucket
+    async fn delete_objects(
+        &self,
+        bucket: &str,
+        identifiers: Vec<ObjectIdentifier>,
+    ) -> Result<DeleteObjectsOutput, RusotoError<DeleteObjectsError>> {
+        let req = DeleteObjectsRequest {
+            bucket: bucket.to_string(),
+            delete: Delete {
+                objects: identifiers,
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+
+        self.s3_client.delete_objects(req).await
+    }
+
+    /// Return all objects keys for the given bucket.
+    async fn list_objects(
+        // the response returns 1000 keys max.
+        // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#listObjectsV2-property
+        &self,
+        bucket: &str,
+        continuation_token: Option<String>,
+    ) -> Result<ListObjectsV2Output, RusotoError<ListObjectsV2Error>> {
+        // list all objects
+        let req = ListObjectsV2Request {
+            bucket: bucket.to_string(),
+            continuation_token,
+            // Minio is not limited to 1000
+            max_keys: Some(1000),
+            ..Default::default()
+        };
+
+        self.s3_client.list_objects_v2(req).await
+    }
+
+    // Unpack the next_continuation_token of the ListObjectsV2Output response
+    fn unpack_next_continuation_token(
+        &self,
+        list_obj_resp: &ListObjectsV2Output,
+    ) -> Option<String> {
+        // https://docs.aws.amazon.com/AmazonS3/latest/dev/ListingObjectKeysUsingJava.html
+        if let Some(is_truncated) = list_obj_resp.is_truncated {
+            if is_truncated {
+                list_obj_resp.next_continuation_token.clone()
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+}
+
+// experimental stream implementation of the list object endpoint
+struct ListObjectIdentifierFuture(
+    Pin<
+        Box<
+            dyn Future<Output = Result<(Option<Vec<ObjectIdentifier>>, Option<String>), ()>> + Send,
+        >,
+    >,
+);
+
+impl Future for ListObjectIdentifierFuture {
+    type Output = Result<(Option<Vec<ObjectIdentifier>>, Option<String>), ()>;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        self.get_mut().0.as_mut().poll(cx)
+    }
+}
+
+impl ListObjectIdentifierFuture {
+    fn new(s3_client: S3Client, bucket: String, continuation_token: Option<String>) -> Self {
+        Self(Box::pin(async move {
+            let req = ListObjectsV2Request {
+                bucket: bucket.clone(),
+                continuation_token: continuation_token.clone(),
+                max_keys: Some(10),
+                ..Default::default()
+            };
+
+            let list_obj_resp = s3_client.list_objects_v2(req).await.map_err(|_| ())?;
+
+            let next_continuation_token = {
```

Comment:
It seems that we already have some helpers to ease the continuation token management (`unpack_next_continuation_token` for instance). Is something preventing us from using them here as well?

Reply:
I had some issues putting it into the impl block of `ListObjectIdentifierFuture`. I can't remember the error. I will give it another try